scipy/stats/_binned_statistic.py

```diff
@@ -664,7 +691,10 @@ def _calc_binned_statistic(Vdim, bin_numbers, result, values, stat_func):
         bin_map = _create_binned_data(bin_numbers, unique_bin_numbers,
                                       values, vv)
         for i in unique_bin_numbers:
-            result[vv, i] = stat_func(np.array(bin_map[i]))
+            stat = stat_func(np.array(bin_map[i]))
+            if np.iscomplexobj(stat) and not np.iscomplexobj(result):
+                raise ValueError("The statistic function returns complex ")
```

Comment:
I think this error message is a bit ambiguous. If we are saying that we support complex numbers, an error that says "The statistic function returns complex " would be confusing for a user. I would recommend changing it to something more descriptive, e.g.:

```suggestion
                raise ValueError("The statistic function returned complex "
                                 f"but `{result.dtype}` was expected. Make "
                                 "sure the function returns the same dtype "
                                 "for empty inputs (`statistic([])`) and "
                                 "non-empty inputs.")
```

Also what about the case where empty input returns a complex value and non-empty input returns non-complex values? i.e. `result` would be complex and `stat` would be non-complex. It makes sense to just coerce to complex but that might be unexpected for the user. I don't think that this case is as critical but, if you are up for a challenge, feel free to address it. Although, I would be more inclined to fix that in a follow-up. I don't want this to stall support for complex numbers.

Reply:
@tirthasheshpatel yes, we should coerce the output to complex numbers. We chose the output `dtype` not knowing what type the user's statistic would produce. Had we known it was complex, we would have allocated a complex array to begin with. 

So I think this idea of trying to use the real array and then falling back to a complex array is right, but I don't know that we want to edit `_calc_binned_statistic`. Instead, how about we set the filter so that [`ComplexWarning`s get raised](https://stackoverflow.com/questions/54814133/catch-numpy-complexwarning-as-exception) and we can catch the `ComplexWarning`?

```python3
import numpy as np
x = np.random.rand(10)
x[:] = 1j*x  # ComplexWarning: Casting complex values to real discards the imaginary part
```

Reply:
Oh, OK. Thanks for the clarifications @mdhaber @cesaregarza! Makes sense to switch to complex numbers and try again. So, no need to change anything!